### DISCUSSION TITLE
Easier batching of bulk inserts

### DISCUSSION THREAD

**Discussion by SebastienGllmt** [SeaQL/sea-orm#710] [Ideas]
## Summary

Postgres only allows a maximum of u16::MAX input parameters. If you exceed this amount, you will get an error in sqlx. Getting around this limitation while using entities is kind of ugly

## Motivation

sea-orm provides a function on entities called `insert_many` such as the the following example

```rust
let apple = cake::ActiveModel {
    name: Set("Apple Pie".to_owned()),
    height: Set(0.5),
    ..Default::default()
};

let insert_result = cake::Entity::insert_many(vec![apple])
     .exec(&db)
     .await?;
```

This insertion requires two SQL parameters because insert_many uses the `VALUES` syntax
```sql
INSERT INTO `cake` (`name`) VALUES (?), (?)
```

This works fine up until you hit the u16 limit of parameters. The naive workaround for the Postgres limit would be to do something like this

```rust
let inserts = vec![apple];
for chunk in inserts.chunks(u16::MAX as usize) {
    cake::Entity::insert_many(chunk.to_vec()).exec(&db).await?;
}
```

However, this won't work because we have to divide by the number of columns! 

That means the proper solution (as far as I can tell) ends up being

```rust
use entity::sea_orm::Iterable;

for chunk in transactions
        .chunks((u16::MAX / <cake::Entity as EntityTrait>::Column::iter().count() as u16) as usize)
    {
    cake::Entity::insert_many(chunk.to_vec()).exec(&db).await?;
}
```

This is already kind of ugly for a one-time patch in your codebase, but having this kind of chunk logic all over the place makes things even worse

It would be nice if insert_many either handled this for you or if there was some other utility function for this.

**Comment by billy1624** [comment]
Hey @SebastienGllmt, thanks for the suggestions! I think we can have a `insert_many_chunk` method in `EntityTrait` which takes an iterator of `ActiveModel` and the chunk size.

Basically, it will perform... for you

```rs
for chunk in inserts.chunks(chunk_size) {
    Self::insert_many(chunk.to_vec()).exec(&db).await?;
}
```

Regarding parameters limit with number of column take into account, it's db dependent and it should be provided by the user. User can substitute it in the batch size just like what you have done.

PR are welcomed! :P

**Reply by SebastienGllmt** [reply]
Hmm, if sea-orm isn't dealing with dividing by the number of columns for you depending on the database you're using, I'm not sure if there is much use in adding this feature to sea-orm. if the function is just a wrapper for `.chunks` the user may as well do that on the user-side.

**Reply by rustonaut** [reply]
I think the chunking should (if possible) be handled  internally and fully opaque to the caller. A `insert_many` without such functionality is a huge foot gun (but a  common foot gun).

Needing chunking is a fundamental limitation/aspect of many-value insert statements(1) no matter the db/binding/orm (The limit for Postgres and MySql might be 65535 binds but for SQLite it is only 32766 and had been 999 in the past. For  MSSQL it is 2100 in case sqlx supports it again in the future.)

_The important thing for such a functionality is that the number of chunks are not a input which should be provided by the caller but one which should  be automatically derived from the query statement based on the number of binds per column._

IMHO if for implementation reasons `insert_many` can't provide this for whatever reason there should be an alternative method which does provide it.

**Reply by tomellm** [reply]
Not sure if this Topic is stale or has already been implemented but I'm of the same opinion. I've already shot myself in the foot twice, and would rather not do it again. Tbh doing the chunking automatically and then allowing for a opt out seems like the better way, all in the name of "sensible defaults". But in any case, since this Discussion board is so old, I'm not sure what the state is. 
Tom ;)

**Reply by Expurple** [reply]
@tomellm, nothing has changed in this regard and the problem is still there.

There have been some good discussions with potential solutions. I think, the best solution for Postgres is `UNNEST ARRAY`: https://github.com/SeaQL/sea-orm/discussions/710#discussioncomment-7319840. MySQL and SQLite can do a chunked insert in a transaction (savepoint): https://github.com/SeaQL/sea-orm/discussions/710#discussioncomment-8445083.

We can do this implicitly by default. But also we're currently preparing the next major release, so in theory we could use it and add a mandatory enum parameter that configures this behavior. I've edited https://github.com/SeaQL/sea-orm/discussions/2548 and listed this as a possibility.

Contributions are welcome!

**Reply by tomellm** [reply]
Hey @Expurple, thanks for the info. I would actually be interested in trying to contribute, but since my schedule is already quite packed I'm not sure if ill have the time. When I do end up getting around to it I would definitely contact you if that's okay ;)

**Comment by tyt2y3** [comment]
I imagine such an API to be like `insert_many().chunked(10)` returning an `Iterator<Item = InsertStatement>` where the user is responsible for executing them.

PR welcomed. Edit: the `10` shall means number of items per chunk.

**Comment by zdcthomas** [comment]
I ran into the same problem this afternoon. I'm considering just going with a less performant approach and just chunking on a low number but I thought I'd check in here to see if anyone's made any progress toward a more elegant solution

**Comment by gitmalong** [comment]
I don't consider the suggested solutions as performant. When thinking about batch inserts remember that inserting 20k rows can be as fast as inserting 10k rows. Hence you don't want to waste time executing the query in relatively small chunks.

I think this problem would not occur if seaorm translates batch inserts to an insert with unnest and array.

```psql
insert into table(a, b) 
values(1, unnest(array[1, 2, 3]));
```

**Reply by rustonaut** [reply]
For context: Example with multiple columns of different type:

```
select * from unnest(ARRAY[1,2,3], ARRAY['a','b','c']);
```

**Reply by Expurple** [reply]
More discussion (and code examples) of this approach with `UNNEST` and `ARRAY`: https://github.com/SeaQL/sea-orm/issues/1922

**Comment by tyt2y3** [comment]
Using `array` is certainly better since we can just put in 1 placeholder.

**Reply by rustonaut** [reply]
Probably the best solution for postgres.

But AFIK it doesn't work (well) with Sqlite and MySql which have the same issue when it comes to `insert_many`.

**Comment by rustonaut** [comment]
One edge case which needs careful consideration in case of any function using internal chunking:

What happens if the call is not run in a transaction but is split into chunks?

a. Should an implicit transaction be started to keep up the behavior that inserts normally all succeed or all fail?
b. Or should be be documented that if used outside of transactions with very a high number of rows*columns 

Not an issue with PG if PG uses the array(s) + unnest approach.

**Reply by Expurple** [reply]
Good catch! I think, on MySQL and SQLite, we should **always** implicitly wrap the chunked insert in a transaction (savepoint). This whole thread is about eliminating unexpected errors and problems. A partially-executed insert would be an even worse gotcha than the original error. And that's the case even when the user code already uses a transaction, because the resulting state that the transaction sees will be unexpected (after calling this method and seeing a database error, I'd expect to see no new rows in the databaase).
